tests/docker/Dockerfile

```diff
@@ -0,0 +1,98 @@
+#
+# This file is part of mbed TLS (https://tls.mbed.org)
+#
+# Copyright (c) 2018, ARM Limited, All Rights Reserved
+#
+# Purpose
+#
+# Defines a Docker container suitable to build and run all tests (all.sh),
+# except for those that use a proprietary toolchain.
+
+FROM ubuntu:xenial
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN apt update \
+    && apt -y install software-properties-common \
+    && rm -rf /var/lib/apt/lists
+
+RUN add-apt-repository -y ppa:team-gcc-arm-embedded/ppa
+
+RUN apt update \
+    && apt -y install \
+    # mbedtls build/test dependencies
+    build-essential \
+    clang \
+    cmake \
+    doxygen \
+    gcc-arm-none-eabi \
+    gcc-mingw-w64-i686 \
+    gcc-multilib \
+    g++-multilib \
+    gdb \
+    git \
+    gnutls-bin \
+    graphviz \
+    lsof \
+    python \
+    python-pip \
+    python3 \
+    wget \
+    # libnettle build dependencies
+    libgmp-dev \
+    m4 \
+    pkg-config \
+    && rm -rf /var/lib/apt/lists/*
+
+# Build a static, legacy openssl from sources with sslv3 enabled
+# Based on https://gist.github.com/bmaupin/8caca3a1e8c3c5686141 (build-openssl.sh)
+RUN cd /tmp \
```

Comment:
I'd like to suggest another OpenSSL versioning scheme. Based on our internally used Dockerfile, I suggest approach like:

    RUN wget -q https://www.openssl.org/source/old/1.0.2/openssl-1.0.2g.tar.gz && \
        tar -zxf openssl-1.0.2g.tar.gz && cd openssl-1.0.2g && \
        ./config --openssldir=/usr/local/openssl-1.0.2g && \
        make clean && make && make install && cd .. && \
        rm -rf openssl-1.0.2g*

I don't mind piping the archive straight into `tar` - that is neat, but please note the installation path; by using only the `--openssldir` configuration argument we get the library installed in a directory named with a proper version number. I find this cleaner than just copying the binary with a new name, straight into `/usr/local/bin`.

You can then follow the command above with:

    ENV OPENSSL=/usr/local/openssl-1.0.2g/bin/openssl

...and have an environmental variable ready for use with Mbed TLS test scripts (as they look for it if you don't provide it with the CLI arguments).

It would be great if you could do it for both the OpenSSL versions.

Reply:
Done. Turns out it's actually `--prefix` that controls where the binary goes (`--openssldir` is for the data), and `make clean` seems not needed given the directory is new. Applied the same for the legacy gnutls.

Reply:
> Turns out it's actually `--prefix` that controls where the binary goes (`--openssldir` is for the data)

Correction: it depends on the OpenSSL version. `openssldir` is correct for 1.0.2g.